packages/datetime/src/dateRangeInput.tsx

```diff
@@ -256,14 +276,219 @@ export class DateRangeInput extends AbstractComponent<IDateRangeInputProps, IDat
+                    isStartInputFocused = true;
+                    isEndInputFocused = false;
+                } else {
+                    // we'd be deselecting the end date on click
+                    isStartInputFocused = false;
+                    isEndInputFocused = true;
+                }
+            } else if (isHoveredEndDefined) {
+                if (isModifyingStartBoundary) {
+                    // we'd be deselecting the start date on click
+                    isStartInputFocused = true;
+                    isEndInputFocused = false;
+                } else {
+                    // we'd be specifying a new end date (clearing the start date) on click
+                    isStartInputFocused = false;
+                    isEndInputFocused = true;
+                }
+            }
+        } else if (isStartDateSelected) {
+            if (isHoveredStartDefined && isHoveredEndDefined) {
+                if (this.areSameDay(hoveredStart, selectedStart)) {
+                    // we'd be modifying the end date on click, so focus the end field
+                    isStartInputFocused = false;
+                    isEndInputFocused = true;
+                } else if (this.areSameDay(hoveredEnd, selectedStart)) {
+                    // we'd be modifying the start date on click, so focus the start field
+                    isStartInputFocused = true;
+                    isEndInputFocused = false;
+                }
+            } else if (isHoveredStartDefined) {
+                // we'd be replacing the start date on click
+                isStartInputFocused = true;
+                isEndInputFocused = false;
+            } else if (isHoveredEndDefined) {
+                // we'd be converting the selected start date to an end date
+                isStartInputFocused = false;
+                isEndInputFocused = true;
+            }
+        } else if (isEndDateSelected) {
+            if (isHoveredStartDefined && isHoveredEndDefined) {
+                if (this.areSameDay(hoveredEnd, selectedEnd)) {
+                    // we'd be modifying the start date on click
+                    isStartInputFocused = true;
+                    isEndInputFocused = false;
+                } else if (this.areSameDay(hoveredStart, selectedEnd)) {
+                    // we'd be modifying the end date on click
+                    isStartInputFocused = false;
+                    isEndInputFocused = true;
+                }
+            } else if (isHoveredEndDefined) {
+                // we'd be replacing the end date on click
+                isStartInputFocused = false;
+                isEndInputFocused = true;
+            } else if (isHoveredStartDefined) {
+                // we'd be converting the selected end date to a start date
+                isStartInputFocused = true;
+                isEndInputFocused = false;
+            }
+        }
```

Comment:
noting that this logic is frightening 😱 
"the cyclomatic complexity is off the charts"